swift/logging/SwiftDiagnostics.h

```diff
@@ -34,7 +34,8 @@ struct SwiftDiagnosticsLocation {
 // Models a diagnostic source for Swift, holding static information that goes out into a diagnostic
 // These are internally stored into a map on id's. A specific error log can use binlog's category
 // as id, which will then be used to recover the diagnostic source while dumping.
-struct SwiftDiagnostic {
+class Diagnostic {
```

Comment:
I think the change from struct to class+constructor is a regression here for two reasons:
 1. Readability at construction time. The designators we had before made it clear what each string argument represents, and we've lost that now. (We could bring it back with the options-struct pattern by adding another struct to hold the constructor parameters, and calling it like `Diagnostic({.id = ...})`, but that feels like overkill.)
 2. Lifetime warnings. Not sure how much this matters since we only expect to create `constexpr` objects, but indirection through the constructor means we [no longer get](https://gcc.godbolt.org/z/5rcPjrr1c) a compiler warning when passing a temporary `std::string` for the parameter.

I'd rather just have `.extractorName = "swift"` throughout, or put it in a global `constexpr std::string_view swiftExtractorName = "swift"`. Or we could even do some trickery like

```
// Diagnostics.h
struct Diagnostic {
 std::string_view extractorName = globalExtractorName;
 ...
};

// SwiftDiagnostics.h
constexpr std::string_view globalExtractorName = "swift";
#include "Diagnostics.h"

// CppDiagnostics.h
constexpr std::string_view globalExtractorName = "cpp";
#include "Diagnostics.h"
```

Reply:
I didn't want to refactor things too much, but I would have avoided `std::string_view` and `constexpr` as being completely unnecessary here, and just stuck with `std::string`. Performance impact would be negligible because diagnostics are not emitted very often. I don't think `std::string_view` is suitable as a field type due to lifetime issues.

I'm not a huge fan of global variables and so `Diagnostics` is better encapsulated. But how about removing the `extractorName` field entirely so that it works the same as `programName` (a global constant). I'm going to try that now and see how it looks.